Update for shell best-practice / portability

charles-dyfis-net opened the pull request:
- Avoid external (OS-vendor-provided) `which` command in favor of shell-builtin and POSIX-defined `command -v` syntax. See also shellcheck warning [SC2230](https://www.shellcheck.net/wiki/SC2230).
- Avoid `echo -n` and `echo -e`; POSIX [does not specify](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/echo.html) either (except to say that output is undefined when the former is used), and whether bash supports them depends on both compile-time flags and runtime flags that can be modified based on environment variable contents at startup time.
- Avoid [legacy ksh function declaration syntax](https://wiki.bash-hackers.org/scripting/obsolete). In ksh (but not bash), `function funcname {` modifies whether variables declared in a function are local; in bash, it has no runtime difference whatsoever from the standardized `funcname() {` syntax.
- Avoid using `==` inside of `[`; the only [POSIX-defined](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/test.html) string comparison operator is `=`.
- Use `$PWD` rather than the much slower `$(pwd)` construct.

Unaddressed in this PR: Using all-caps variable names ([namespace reserved](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html) for variables meaningful to the operating system and shell -- when reading the linked specification, keep in mind that modifying a shell variable updates any like-named environment variable), and use of `errexit` (which has [deeply unintuitive](http://mywiki.wooledge.org/BashFAQ/105#Exercises) and [nonportable](https://www.in-ulm.de/~mascheck/various/set-e/) behavior; explicit error handling is more consistent and thus easier to audit).

ideaship commented on ttar:
```diff
@@ -28,7 +28,7 @@ export LC_ALL=C
 path=""
 CMD=""
-ARG_STRING="$*"
+printf -v ARG_STRING '%q ' "$@"
```
run_test.sh fails due to this change: the ARG_STRING now has a trailing space. We need to fix that or update the test.

charles-dyfis-net replied:
Can follow up with `ARG_STRING=${ARG_STRING% }` as a following command to trim that.

ideaship reviewed (changes requested):
First of all, thank you for the patch and sorry for not getting back to you earlier.

Do you have a tool that flags POSIX issues or is this just based on your knowledge?

xplshn commented:
> First of all, thank you for the patch and sorry for not getting back to you earlier.
> 
> Do you have a tool that flags POSIX issues or is this just based on your knowledge?

`shellcheck` will print all of these errors, warnings and more

charles-dyfis-net committed:
Update for shell best-practice / portability

- Avoid `echo -n` and `echo -e`; POSIX does not specify either (except to say
  that output is undefined when the former is used), and whether bash supports
  them depends on both compile-time flags and runtime flags that can be
  modified based on environment variable contents at startup time.

- Avoid legacy ksh function declaration syntax. In ksh (but not bash),
  `function funcname {` modifies whether variables declared in a function are
  local; in bash, it has no runtime difference whatsoever from the standardized
  `funcname() {` syntax.

- Avoid using `==` inside of `[`; the only POSIX-defined string comparison
  operator is `=`.

- Use `$PWD` rather than the much slower `$(pwd)` construct.

- Use the regular `python` command unless the environment variable `PYTHON` is
  set **and** no exported function is already overriding `python`.

ideaship commented:
Woah, blast from the past! Thanks for contributing to this project.
